### PR TITLE
8261433: Better pkcs11 performance for libpkcs11:C_EncryptInit/libpkcs11:C_DecryptInit

### PR DESCRIPTION
**Pull request by pkumaraswamy**
This fix intends to eliminate additional library call to C_EncryptInit or C_DecryptInit for Ciphers running through the CKM_AES_GCM.

Background: 

There are two types of CK_GCM_PARAMS struct that are used, one with IV bits and the other without it.

Initially there was issue in NSS library, due to the struct being different in header and spec version.
NSS was using version from header but Solaris and SoftHsm was using normative version from spec.
To maintain compatibility Java used to try library call with non-normative (header) version first and then upon failure retrial was made with updated GCM struct with IV bits.

Note: Trying normative (spec) version first with NSS library results in JVM crash.

Refer below for more information:
https://github.com/openjdk/jdk/blob/master/src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11gcm2.h#L36  

However NSS has fixed this to use normative/spec version since 3.52 which has spec version 2.40
Solaris and SoftHSM was already complying to the version mentioned in spec 2.40

The fix now check if spec version is 2.40 and then makes library call with appropriate structure.

Internal testing is green, further I have done internal testing manually with NSS library 3.96, 3.76, 3.51 (non-normative spec), 3.52 and 3.53
Results are attached [nss_logs.zip](https://github.com/openjdk/jdk/files/14692787/nss_logs.zip)

Our existing tests like sun/security/pkcs11/Cipher/TestKATForGCM.java already tests the functionality and I have used the same for internal testing

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8261433](https://bugs.openjdk.org/browse/JDK-8261433): Better pkcs11 performance for libpkcs11:C_EncryptInit/libpkcs11:C_DecryptInit (**Bug** - P3)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to [312859f5](https://git.openjdk.org/jdk/pull/18425/files/312859f536f3836d5cdb3f6d5358741cc8cdaa60)
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18425/head:pull/18425` \
`$ git checkout pull/18425`

Update a local copy of the PR: \
`$ git checkout pull/18425` \
`$ git pull https://git.openjdk.org/jdk.git pull/18425/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18425`

View PR using the GUI difftool: \
`$ git pr show -t 18425`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18425.diff">https://git.openjdk.org/jdk/pull/18425.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18425#issuecomment-2011741062)

### CODE REVIEW

**Comment by valeriepeng** [comment]
Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,

1.  update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead. 
2.  enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.

**Comment by pkumaraswamy** [comment]
> Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,
> 
> 1. update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead.
> 2. enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.

If PKCS11 spec version is less than 2.40, we are retaining the same logic as it exists today. 
If we make the above changes then there will be a crash in NSS version 3.51 and below where PKCS11 2.20 spec is used.

However, I can optimze the code by introducing new boolean variable in C_DecryptInit and C_EncryptInit to indicate if we need to send normative struct. If so, I'll add a goto statement to updateGCM params.
I did not do this initially because in 11u and above I have seen goto being removed.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java [hunk 1]
```diff
@@ -454,12 +454,32 @@ private void initialize() throws PKCS11Exception {
...
                 session = token.getOpSession();
             }
+
+            /*
+             * Check if PKCS11 spec version is 2.40 and above.
+             * If so, we need to send CK_GCM_PARAMS structure with IV bits for AES GCM.
+             * the call to C_EncryptInitWithIvBitsMech will update GCM structure, with
+             * additional field ulvIV bits and then invoke C_EncryptInit
+             */
+            CK_VERSION cryptokiVersion = token.p11.C_GetInfo().cryptokiVersion;
```
Comment (djelinski) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java:464]: use p11.getVersion() instead

**Review by djelinski — COMMENTED** [review]
Does this PR actually improve the performance of AES/GCM? I'm asking because NSS documentation states that the newer NSS versions actually support both forms of the parameters, so I'd expect no performance penalty for using the old version. Is this on some other provider?

**Comment by valeriepeng** [comment]
> Does this PR actually improve the performance of AES/GCM? I'm asking because NSS documentation states that the newer NSS versions actually support both forms of the parameters, so I'd expect no performance penalty for using the old version. Is this on some other provider?

Yes, IIRC, the performance improvement is meant for non-NSS providers which only accepts the normative version. For these providers, the first call would fail due to invalid parameters, thus requiring 2 calls. This is due to earlier NSS impl would crash when given normative version. Thus the ordering of trying non-normative version before normative version.

**Comment by valeriepeng** [comment]
> > Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,
> > 
> > 1. update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead.
> > 2. enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.
> 
> If PKCS11 spec version is less than 2.40, we are retaining the same logic as it exists today. If we make the above changes then there will be a crash in NSS version 3.51 and below where PKCS11 2.20 spec is used.
> 
> However, I can optimze the code by introducing new boolean variable in C_DecryptInit and C_EncryptInit to indicate if we need to send normative struct. If so, I'll add a goto statement to updateGCM params. I did not do this initially because in 11u and above I have seen goto being removed.

Yes, it's better to add one boolean and just have one EncryptInit/DecryptInit() method. As times goes by, I'd expect old NSS library usage to die out and we should optimize the flow toward normative version.

**Comment by pkumaraswamy** [comment]
> > > Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,
> > > 
> > > 1. update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead.
> > > 2. enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.
> > 
> > 
> > If PKCS11 spec version is less than 2.40, we are retaining the same logic as it exists today. If we make the above changes then there will be a crash in NSS version 3.51 and below where PKCS11 2.20 spec is used.
> > However, I can optimze the code by introducing new boolean variable in C_DecryptInit and C_EncryptInit to indicate if we need to send normative struct. If so, I'll add a goto statement to updateGCM params. I did not do this initially because in 11u and above I have seen goto being removed.
> 
> Yes, it's better to add one boolean argument and just have one EncryptInit/DecryptInit() method. As times goes by, I'd expect old NSS library usage to die out and we should optimize the flow toward normative version.

I just noticed, If I add boolean variable, it'll affect all classes (like P11RSA.java etc.) that extends PKCS11.java and we will have to modify all subclasses.

**Comment by valeriepeng** [comment]
> > > > Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,
> > > > 
> > > > 1. update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead.
> > > > 2. enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.
> > > 
> > > 
> > > If PKCS11 spec version is less than 2.40, we are retaining the same logic as it exists today. If we make the above changes then there will be a crash in NSS version 3.51 and below where PKCS11 2.20 spec is used.
> > > However, I can optimze the code by introducing new boolean variable in C_DecryptInit and C_EncryptInit to indicate if we need to send normative struct. If so, I'll add a goto statement to updateGCM params. I did not do this initially because in 11u and above I have seen goto being removed.
> > 
> > 
> > Yes, it's better to add one boolean argument and just have one EncryptInit/DecryptInit() method. As times goes by, I'd expect old NSS library usage to die out and we should optimize the flow toward normative version.
> 
> I just noticed, If I add boolean variable, it'll affect all classes (like P11RSA.java etc.) that extends PKCS11.java and we will have to modify all subclasses which might not be the right approach.

Hmm, good point, let me try it out and see if there are other choices.

**Comment by valeriepeng** [comment]
> > Now that we are going with the normative version first, maybe we should make additional changes to clean up the flow further?Say,
> > 
> > 1. update `jGCMParamsToCKGCMParamPtr(JNIEnv *env, jobject jParam, CK_ULONG *pLength)` to allocate the normative structure instead.
> > 2. enhance `updateGCMParams(JNIEnv *env, CK_MECHANISM_PTR mechPtr)` to return a copy of mech pointer containing the non-normative structure.
> 
> If PKCS11 spec version is less than 2.40, we are retaining the same logic as it exists today. If we make the above changes then there will be a crash in NSS version 3.51 and below where PKCS11 2.20 spec is used.
> 
> However, I can optimze the code by introducing new boolean variable in C_DecryptInit and C_EncryptInit to indicate if we need to send normative struct. If so, I'll add a goto statement to updateGCM params. I did not do this initially because in 11u and above I have seen goto being removed.

I prototyped this a bit and we can add GCM specific encrypt/decrypt init methods which takes additional boolean argument which indicate which version to use first for the PKCS11 init call. This way, we can remove the mechanism check in C_EncryptInit/C_DecryptInit since GCM init would go through the new GCM specific encrypt/decrypt init methods. I ended up also making changes to p11_convert.c and p11_util.c as well to use the normative version by default. 
It's easier to explain though the webrev/patch which I will share with you through slack. I don't have the older NSS library to verify the prototype changes, but it works fine w/ the current NSS library in the artifactory. Maybe you can take a look and let me know what do you think. Thanks!

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java [hunk 2]
```diff
@@ -793,6 +793,24 @@ public native long[] C_FindObjects(long hSession, long ulMaxObjectCount)
...
+     * with alternative param structure for max compatibility.
+     * (Encryption and decryption)
+     *
+     * @param hSession the session's handle
+     *         (PKCS#11 param: CK_SESSION_HANDLE hSession)
+     * @param pMechanism the encryption mechanism
+     *         (PKCS#11 param: CK_MECHANISM_PTR pMechanism)
+     * @param hKey the handle of the encryption key
+     *         (PKCS#11 param: CK_OBJECT_HANDLE hKey)
+     * @param useNormativeVerFirst which version of params to use first
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java:807]: nit: maybe "whether to use normative version of GCM parameters first"?

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java [hunk 3]
```diff
@@ -793,6 +793,24 @@ public native long[] C_FindObjects(long hSession, long ulMaxObjectCount)
     public native void C_EncryptInit(long hSession, CK_MECHANISM pMechanism,
             long hKey) throws PKCS11Exception;
 
+    /**
+     * C_GCMEncryptInitWithRetry initializes an GCM encryption operation and retry
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java:797]: nit: an -> a?

**Comment by valeriepeng** [comment]
Please make sure the testing covers old and new NSS versions just to be safe.
Thanks!

**Review by valeriepeng — APPROVED** [review]

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java (hunk 1 above)
Comment (pkumaraswamy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java]: Done

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java (hunk 2 above)
Comment (pkumaraswamy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java]: Done

**Comment by pkumaraswamy** [comment]
> Please make sure the testing covers old and new NSS versions just to be safe. Thanks!

Done, I have attached test results of old and new NSS, However I have also tested few other NSS between 3.46 and the latest one
[nsstest.zip](https://github.com/openjdk/jdk/files/15272353/nsstest.zip)

**Review by djelinski — APPROVED** [review]

**Review by coffeys — APPROVED** [review]

**Pull request closed without merging**